submissions/dashakim/oop-task/js/app.js

```diff
@@ -0,0 +1,125 @@
+var gameScore = 0;
+
+const ENEMY = {
+  speed: 100,
+  varianceSpeed: 222,
+  location: [70, 155, 235]
+};
+const PLAYER = {
+  startX: 200,
+  startY: 404,
+  stepX: 101,
+  stepY: 83
+};
+
+const CANVAS = {
+  begin: -60,
+  end: 510,
+  leftSide: 0,
+  rightSide: 400,
+  top: 0,
+  bottom: 404,
+  halfCell: 50
+};
+
+var Character = function(x, y, sprite) {
+  this.x = x;
+  this.y = y;
+  this.sprite = sprite;
+};
+Character.prototype.constructor = Character;
+Character.prototype.render = function() {
+  ctx.drawImage(Resources.get(this.sprite), this.x, this.y);
+};
+
+var Enemy = function(x, y, speed, player) {
+  Character.call(this, x, y);
+  this.sprite = "images/car.png";
+  this.speed = speed;
+  this.player = player;
+};
+Enemy.prototype = Object.create(Character.prototype);
+Enemy.prototype.constructor = Enemy;
+Enemy.prototype.toRender = function() {
+  this.render();
+};
+
+Enemy.prototype.update = function(dt) {
+  this.x += this.speed * dt;
+  if (this.x > CANVAS.end) {
+    this.x = CANVAS.begin;
+    //random speed for enemies
+    this.speed = ENEMY.speed + Math.floor(Math.random() * ENEMY.varianceSpeed);
+  }
+  // return to start position player when collide with enemy
+  if (
+    this.player.x > this.x - CANVAS.halfCell &&
+    this.player.x < this.x + CANVAS.halfCell &&
+    this.player.y > this.y - CANVAS.halfCell &&
+    this.player.y < this.y + CANVAS.halfCell
+  ) {
+    this.player.goToStart();
+  }
+};
+
+var Player = function(x, y, addScore) {
+  Character.call(this, x, y);
+  this.addScore = addScore;
+  this.sprite = "images/char-cat-girl.png";
+};
+
+Player.prototype = Object.create(Character.prototype);
+Player.prototype.constructor = Player;
+Player.prototype.toRender = function() {
+  this.render();
+};
+
+Player.prototype.update = function() {
+  if (this.y <= 0) {
+    this.addScore();
+    this.goToStart();
+  }
+};
+
+Player.prototype.goToStart = function() {
+  this.x = PLAYER.startX;
+  this.y = PLAYER.startY;
+};
+
+Player.prototype.handleInput = function(keyUp) {
+  if (keyUp === "left" && this.x > CANVAS.leftSide) {
+    this.x = this.x - PLAYER.stepX;
+  } else if (keyUp === "right" && this.x < CANVAS.rightSide) {
+    this.x = this.x + PLAYER.stepX;
+  } else if (keyUp === "up" && this.y > CANVAS.top) {
+    this.y = this.y - PLAYER.stepY;
+  } else if (keyUp === "down" && this.y < CANVAS.bottom) {
+    this.y = this.y + PLAYER.stepY;
+  }
+};
+
+function addScoreBy() {
+  return function() {
+    return gameScore++;
+  };
+}
+var count = addScoreBy();
+
+var displayStats = function() {
+  count();
```

Comment:
Why introducing `count`? The question is not rhetoric. Need to understand your way of thinking.

Reply:
I just wanted to call function addScoreBy. And I'm back to  the previous version

Reply:
This is ok. Why not directly but via `count`?

Reply:
hmm, if I will call addScoreBy() I will get [function], or you mean I should write:
function addScoreBy() {
      return gameScore++;
    };
addScoreBy()

Reply:
Try simplifying this fragment as much as possible (including returning a function from `addScoreBy()`) and see what happens. Add complexity and abstraction only when really in need

Reply:
BTW,
```
function doSomething() {
}

var a = new WhatEver(doSomething);
```

Is OK. `doSomething` is already a reference.

Please do not mix ES5 with ES6, stick to using `var` in this task.

Reply:
I use const only for constant variable. I thought it's a good practice. I didn't find any "let"
Ok. I think you have had enough of this PR. But one last hint, what actually I should improve?

Reply:
Let's do some clean up. And I will take a fresh look onto the code.